jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java

```diff
@@ -179,25 +178,25 @@ public PendingFutureLimiter changeThresholdFactor(float thresholdFactor) {
      */
     protected <T> CompletableFuture<T> internalEnqueue(CompletableFuture<T> future,
                                                        boolean needToBlock) throws InterruptedException {
-        if (counter.get() == maxPendingCount && thresholdListener != null) {
-            thresholdListener.onHiLimitReached();
-        }
-
         if (needToBlock) {
             awaitOpportunityToEnqueueAndPurge();
         }
 
-        counter.incrementAndGet();
-        pendingFutures.put(future, System.currentTimeMillis());
-        future.handleAsync((any, exc) -> {
+        CompletableFuture<T> resultFuture = future.handleAsync((any, exc) -> {
             if (exc != null) {
                 log.error(exc.getMessage(), exc);
+                throw new CompletionException(exc);
+            } else {
+                return any;
             }
+        });
+
+        resultFuture.handleAsync((any, exc) -> {
             long value = counter.decrementAndGet();
-            pendingFutures.remove(future);
+            pendingFutures.remove(resultFuture);
 
             if (value == 0 || value == getThreshold()) {
-                if (thresholdListener != null) {
+                if (value == getThreshold() && thresholdListener != null) {
```

Comment:
`changeThresholdFactor` does not allows to disable threshold or to increase it to the size of a maxPendingCount.  
Suggestion: replace
`if (threshold <= 0 || threshold >= maxPendingCount)` 
with
`if (threshold < 0 || threshold > maxPendingCount)`

You can see that currently: 
limiter.changeThresholdFactor(0);
limiter.changeThresholdFactor(1) 
will rise an exception.

Reply:
Changed. 
Actually, right now we can change thresholdFactor to negative or more than 1 values, as long as maxPendingCount is small enough. In this case we'll rise exception only in setMaxPendingCount. I am not sure if this behavior is totally expected... Added a test, illustrating current behavior.

Reply:
Could you please take an effort and provide correct validation to setMaxPendingCount/setThresholdFactor methods so they would behave correctly and user of the library does not blow everything by series of incorrect configurations.

Also  do you REALLY think  that allowing thresholdFactor to be NEGATIVE value is a GOOD Idea? Have you looked at changeThresholdFactor docs?

Reply:
No, I don't think this is a good idea, and that what was my question about. 
There weren't anything about border values in changeThresholdFactor docs, so I have added it.
Changed validation method, updated tests.